pypose/function/spline.py

```diff
@@ -0,0 +1,195 @@
+                u(t)^2\\
+                u(t)^3\\
+            \end{bmatrix}
+
+    where :math:`u(t)=(t-t_i)/\Delta t`, :math:`\Delta t = t_{i+1} - t_{i}`. :math:`t_i`
+    is the time point of the :math:`i_{th}` given pose. :math:`t` is the interpolation
+    time point :math:`t \in [t_{i},t_{i+1})`.
+
+    :math:`\delta \hat{\xi}_j` is the transformation between :math:`\hat{\xi}_{j-1}`
+    and :math:`\hat{\xi}_{j}`
+
+    .. math::
+        \begin{aligned}
+            \delta \hat{\xi}_j :&= \log(\exp(\hat{\xi} _w^{j-1})\exp(\hat{\xi}_j^w))
+                                &= \log(T_j^{j-1})\in \mathfrak{se3}
+        \end{aligned}
+
+    Note:
+        The implementation is based on Eq. (3), (4), (5), and (6) of this paper:
+        David Hug, et al.,
+        `HyperSLAM: A Generic and Modular Approach to Sensor Fusion and Simultaneous
+        Localization And Mapping in Continuous-Time
+        <https://ieeexplore.ieee.org/abstract/document/9320417>`_,
+        2020 International Conference on 3D Vision (3DV), Fukuoka, Japan, 2020.
+
+    Examples:
+        >>> import torch
+        >>> import pypose as pp
+        >>> a1 = pp.euler2SO3(torch.Tensor([0., 0., 0.]))
+        >>> a2 = pp.euler2SO3(torch.Tensor([torch.pi / 4., torch.pi / 3., torch.pi / 2.]))
+        >>> time = torch.arange(0, 1, 0.25).reshape(1, 1, -1)
+        >>> poses = pp.LieTensor([[
+        ...                             [0., 4., 0., a1[0], a1[1], a1[2], a1[3]],
+        ...                             [0., 3., 0., a1[0], a1[1], a1[2], a1[3]],
+        ...                             [0., 2., 0., a1[0], a1[1], a1[2], a1[3]],
+        ...                             [0., 1., 0., a1[0], a1[1], a1[2], a1[3]],
+        ...                             [1., 0., 1., a2[0], a2[1], a2[2], a2[3]],
+        ...                             [2., 0., 1., a2[0], a2[1], a2[2], a2[3]],
+        ...                             [3., 0., 1., a2[0], a2[1], a2[2], a2[3]],
+        ...                             [4., 0., 1., a2[0], a2[1], a2[2], a2[3]]
+        ...                            ]], ltype=pp.SE3_type)
+        >>> wayposes = pp.BSplineSE3(poses, time)
+
+    - .. figure:: /_static/img/module/liespline/BsplineSE3.png
+
+        Fig. 1. Result of B Spline Interpolation in SE3.
+    '''
+    assert is_SE3(data), "The input poses are not SE3Type."
+    assert time.shape[0]==time.shape[1]==1, "The time has wrong shape."
+    data = torch.cat((data[..., :1, :], data, data[..., -1:, :]), dim=-2)
+    K, N = time.shape[-1], data.shape[-2]
+    dargs = {'dtype': time.dtype, 'device': time.device}
+    alpha = torch.arange(4, **dargs).view(-1, 1)
+    tt = time ** alpha
+    B = torch.tensor([[5, 3,-3, 1],
+                      [1, 3, 3,-2],
+                      [0, 0, 0, 1]], **dargs) / 6
+    T_delta = data[..., 0:-3, :].unsqueeze(2).repeat(1, 1, K, 1)
+    w = (B @ tt).unsqueeze(-1)
+    P = torch.stack([data[..., i:i + 4, :] for i in range(N - 3)], dim=1).unsqueeze(-2)
```

Comment:
please try to remove the loop, you may generate the index first without loops, then use the index for data accessing.